### PR TITLE
Improve error message for `assert_column_eq` in pylibcudf tests

### PR DESCRIPTION
**Pull request by TomAugspurger**
## Description

This changes our test-only assert_column_eq utility to provide a better error message when running under pytest.

After we've detected that two columns are not equal, we'll cast the columns to either python lists (for nested dtypes like lists or structs) or numpy arrays.

We then use NumPy's or python's usual assertion machiner to compare.

Some examples:

**Numeric values** (using NumPy)

```pytb
python/pylibcudf/tests/test_utils.py:34: in test_assert_column_eq_ok_raises
    assert_column_eq(column, pa.array(right))
python/pylibcudf/tests/common/utils.py:195: in assert_column_eq
    np.testing.assert_array_equal(left, right)
E   AssertionError: 
E   Arrays are not equal
E   
E   Mismatched elements: 1 / 3 (33.3%)
E   Max absolute difference among violations: 1
E   Max relative difference among violations: 0.25
E    ACTUAL: array([1, 2, 3])
E    DESIRED: array([1, 2, 4])
```

**List-like values** (using python lists)

```pytb
python/pylibcudf/tests/test_utils.py:34: in test_assert_column_eq_ok_raises
    assert_column_eq(column, pa.array(right))
python/pylibcudf/tests/common/utils.py:197: in assert_column_eq
    assert left == right
E   assert [[1, 2], [3]] == [[1, 2], [3, 4]]
E     
E     At index 1 diff: [3] != [3, 4]
E     Use -v to get more diff
```

**Dict-like values**

```pytb
python/pylibcudf/tests/test_utils.py:34: in test_assert_column_eq_ok_raises
    assert_column_eq(column, pa.array(right))
python/pylibcudf/tests/common/utils.py:197: in assert_column_eq
    assert left == right
E   AssertionError: assert [{'a': 1}, {'a': 2}] == [{'a': 1}, {'a': 3}]
E     
E     At index 1 diff: {'a': 2} != {'a': 3}
E     Use -v to get more diff
```

Towards https://github.com/rapidsai/cudf/issues/19346

### CODE REVIEW

**Review by galipremsagar — APPROVED** [review]

File: python/pylibcudf/tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    [
+        ([1, 2, 3], [1, 2, 4]),
+        ([[1, 2], [3]], [[1, 2], [3, 4]]),
+        ([{"a": 1}, {"a": 2}], [{"a": 1}, {"a": 3}]),
+    ],
+)
+def test_assert_column_eq_ok_raises(left: list, right: list) -> None:
+    array = pa.array(left)
+    column = plc.Column.from_arrow(array)
+    with pytest.raises(AssertionError):
```
Comment (Matt711) [inline python/pylibcudf/tests/test_utils.py]: nitpick
```suggestion
    with pytest.raises(AssertionError, match="..."):
```

File: python/pylibcudf/tests/conftest.py [hunk 2]
```diff
@@ -15,6 +15,13 @@
 
 sys.path.insert(0, os.path.join(os.path.dirname(__file__), "common"))
 
+
+# This ensures that assertions from expressions like `left == right` in the
+# common/utils.py are rewritten so that we get nice tracebacks.
+# We need to register this before the module is imported.
+pytest.register_assert_rewrite("utils")
```
Comment (Matt711) [inline python/pylibcudf/tests/conftest.py:22]: cool TIL

File: python/pylibcudf/tests/common/utils.py [hunk 3]
```diff
@@ -183,7 +183,23 @@ def _is_supported_for_pc_is_nan(arr_type):
 
             np.testing.assert_array_almost_equal(lh_arr, rh_arr)
     else:
-        assert lhs.equals(rhs)
+        if not lhs.equals(rhs):
```
Comment (mroeschke) [inline python/pylibcudf/tests/common/utils.py:187]: Just FYI on line 176 there's another `assert lhs_nans.equals(rhs_nans)` that could use the same treatment.

File: python/pylibcudf/tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+from utils import assert_column_eq
+
+import pylibcudf as plc
+
+
+@pytest.mark.parametrize(
+    "values",
+    [
+        [1, 2, 3],
+        [[1, 2], [3]],
```
Comment (mroeschke) [inline python/pylibcudf/tests/test_utils.py:14]: If you need to add another commit, could you add a test with one of the values being `None`?

**Review by mroeschke — APPROVED** [review]

File: python/pylibcudf/tests/test_utils.py (hunk 1 above)
Comment (TomAugspurger) [reply python/pylibcudf/tests/test_utils.py]: Added some matches, but I kept it pretty generic. These are matching messages that are controlled by NumPy (for the case of numeric arrays that can be zero-copy converted from arrow to numpy) or python lists. Since error messages tend to change from release to release, I don't want to be too specific about what we're expecting.
